### PR TITLE
Llama4 unified checkpoint export

### PR DESCRIPTION
**Pull request by yueshen2016**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add support to Llama4 unified checkpoint export after the nemo checkpoint get PTQ'ed.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/modelopt/quantization/quantizer.py [hunk 1]
```diff
@@ -330,8 +331,29 @@ def quantize(self, model: "MegatronParallel", forward_loop=None):
...
-        if is_global_rank_zero():
-            mtq.print_quant_summary(unwrapped_model)
+        # Print quantization summary with both tensor and pipeline parallelism awareness
+        from megatron.core import parallel_state
+
+        if parallel_state.is_initialized():
+            pp_rank = parallel_state.get_pipeline_model_parallel_rank()
+            pp_size = parallel_state.get_pipeline_model_parallel_world_size()
+            tp_rank = parallel_state.get_tensor_model_parallel_rank()
+            tp_size = parallel_state.get_tensor_model_parallel_world_size()
```
Comment (kevalmorabia97) [inline nemo/collections/llm/modelopt/quantization/quantizer.py]: tp_size is not used anywhere

File: nemo/collections/llm/modelopt/quantization/quantizer.py [hunk 2]
```diff
@@ -330,8 +331,29 @@ def quantize(self, model: "MegatronParallel", forward_loop=None):
                 unwrapped_model, "*input_quantizer", lambda amax: torch.clamp(amax, min=0.01 * maxbound)
             )
 
-        if is_global_rank_zero():
-            mtq.print_quant_summary(unwrapped_model)
+        # Print quantization summary with both tensor and pipeline parallelism awareness
+        from megatron.core import parallel_state
```
Comment (kevalmorabia97) [inline nemo/collections/llm/modelopt/quantization/quantizer.py]: we can move the import to the top of this file

File: nemo/collections/llm/modelopt/quantization/quantizer.py [hunk 3]
```diff
@@ -330,8 +331,29 @@ def quantize(self, model: "MegatronParallel", forward_loop=None):
...
+
+        if parallel_state.is_initialized():
```
Comment (kevalmorabia97) [inline nemo/collections/llm/modelopt/quantization/quantizer.py]: Do you see any case where parallel_state is not already initialized in this function call?

File: nemo/collections/llm/modelopt/quantization/quantizer.py [hunk 4]
```diff
@@ -480,12 +506,17 @@ def export_hf_checkpoint(
...
         with torch.inference_mode():
             with tempfile.TemporaryDirectory() as tmp_dir:
                 exporter.config.save_pretrained(tmp_dir)
+                # For llama4, we only deal with the language_model, vision_model and multi_modal_projector will be acquired from the huggingface checkpoint
+                if (
+                    hasattr(unwrapped_model, 'language_model')
+                    and hasattr(unwrapped_model, 'vision_model')
+                    and hf_checkpoint
+                ):
+                    tmp_dir = hf_checkpoint
```
Comment (kevalmorabia97) [inline nemo/collections/llm/modelopt/quantization/quantizer.py]: I guess `exporter.config.save_pretrained(tmp_dir)` is not needed in Llama4 case so lets move it to `else` section.

Also, for VLMs, we should `assert hf_checkpoint is not None`

```suggestion
                # For llama4, we only deal with the language_model. vision_model and multi_modal_projector will be acquired from the huggingface checkpoint
                if (
                    hasattr(unwrapped_model, 'language_model')
                    and hasattr(unwrapped_model, 'vision_model')
                ):
                    assert hf_checkpoint is not None, "hf_checkpoint is required for exporting VLMs!"
                    tmp_dir = hf_checkpoint
                else:
                    exporter.config.save_pretrained(tmp_dir)
```

File: nemo/collections/llm/modelopt/quantization/quantizer.py (hunk 1 above)
Comment (yueshen2016) [reply nemo/collections/llm/modelopt/quantization/quantizer.py]: Have deprecated this variable.

File: nemo/collections/llm/modelopt/quantization/quantizer.py (hunk 3 above)
Comment (yueshen2016) [reply nemo/collections/llm/modelopt/quantization/quantizer.py]: Fixed this.

File: nemo/collections/llm/modelopt/quantization/quantizer.py (hunk 4 above)
Comment (yueshen2016) [reply nemo/collections/llm/modelopt/quantization/quantizer.py]: Fixed, thanks!

**Review by pablo-garay — APPROVED** [review]

**Comment by pablo-garay** [comment]
approved from Automation/tests perspective. Please reach out to further reviewer(s) based on our conversation

**Review by oyilmaz-nvidia — DISMISSED** [review]
LGTM. Just FYI, llm collection is deprecated in this repo and we shouldn't be adding more PRs to this repo unless it's urgent.

**Review by kevalmorabia97 — DISMISSED** [review]
